usb: usdc: BL61x Device HS USB

VynDragon opened the pull request:
It does device at high-speed speeds. 

Caveat: controller does host, this does not do host.

VynDragon committed:
usb: udc: Introduce bl61x UDC driver

Adds a device mode driver for BL61x

Signed-off-by: Camille BAUD <mail@massdriver.space>

VynDragon committed:
dts: Add BL61x USB device binding

Adds the binding to enable the USB controller

Signed-off-by: Camille BAUD <mail@massdriver.space>

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+struct udc_bflb_bl61x_data {
+	struct k_work work;
+	struct k_fifo fifo;
+	bool ep_in[5];
```
If wanting to proactively reduce chance of confusion:
```suggestion
	bool ep_is_in[5];
```
Though the type, bool, makes it obvious already.

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+			group_intstatus &=  ~sys_read32(cfg->base + USB_DEV_MISG0_OFFSET);
+
+			if (group_intstatus & USB_CX_COMABT_INT) {
+				udc_submit_event(dev, UDC_EVT_ERROR, -1);
```
Maybe `-ECANCELED` is fitting?
```suggestion
				udc_submit_event(dev, UDC_EVT_ERROR, -ECANCELED);
```

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+			if (group_intstatus & USB_ISO_SEQ_ERR_INT) {
+				sys_write32(USB_ISO_SEQ_ERR_INT, cfg->base + USB_DEV_ISG2_OFFSET);
+
+				udc_submit_event(dev, UDC_EVT_ERROR, -1);
```
Probably `-EIO` as it tried some I/O and failed.
```suggestion
				udc_submit_event(dev, UDC_EVT_ERROR, -EIO);
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+			if (group_intstatus & USB_ISO_SEQ_ABORT_INT) {
+				sys_write32(USB_ISO_SEQ_ABORT_INT, cfg->base + USB_DEV_ISG2_OFFSET);
+
+				udc_submit_event(dev, UDC_EVT_ERROR, -1);
```
```suggestion
				udc_submit_event(dev, UDC_EVT_ERROR, -ECANCELED);
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+	return 0;
+}
+
+
```
A double newline hiding there
```suggestion
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+	int ret;
+
+	ret = k_mem_slab_alloc(&udc_bflb_bl61x_ev_slab, (void **)&ev, K_NO_WAIT);
+	if (ret) {
```
I think the USB area started to anticipate a possible MISRA compliance, which wants `if (expr)` have `expr` be a bool (seen [here](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html), row 85).

And switching from `ret` to `err` would be fitting the rest of USB codebase. Maybe this would be pointed at for next reviewers.

```suggestion
	int err;

	err = k_mem_slab_alloc(&udc_bflb_bl61x_ev_slab, (void **)&ev, K_NO_WAIT);
	if (err != 0) {
```

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+	lock_key = irq_lock();
+
+	buf = udc_buf_get_all(cfg);
+	if (buf) {
```
I think the USB area started to anticipate a possible MISRA compliance, which wants `if (expr)` have `expr` be a bool (seen [here](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html), row 85).

```suggestion
	if (buf != NULL) {
```

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+	udc_bflb_bl61x_fifo_reset(dev, 3);
+	udc_bflb_bl61x_fifo_reset(dev, 4);
+
+	/* enable 'vdma' (virtual dma? video dma?) */
```
BL808 [says](https://github.com/bouffalolab/bl_iot_sdk/blob/edcf613f073d3ef7d2f290d7c285b69be35da75b/components/platform/soc/bl808/bl808_e907_std/bl808_bsp_driver/std_drv/inc/bl808_usb.h#L222-L229) it's `virtual`

```c
/**
 *  @brief USB Virtual DMA transfer type definition
 */
typedef enum {
    USB_VDMA_TRANS_DIR_FIFO_2_MEM = 0,  /*!< USB VDMA transfer type is FIFO to memory */
    USB_VDMA_TRANS_DIR_MEM_2_FIFO = 1,  /*!< USB VDMA transfer type is memory to FIFO */
    USB_VDMA_TRANS_DIR_FIFO_2_FIFO = 2, /*!< USB VDMA transfer type is FIFO to FIFO */
} USB_VDMA_TRANS_DIR_Type;
```

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+	tmp |= USB_VDMA_EN;
+	sys_write32(tmp, cfg->base + USB_VDMA_CTRL_OFFSET);
+
+	/* force plug signal? */
```
I understand it as "do not force-unplug", where the `|= USB_UNPLUG` would force-eject everything.

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+
+	priv = CONTAINER_OF(item, struct udc_bflb_bl61x_data, work);
+
+	while ((ev = k_fifo_get(&priv->fifo, K_NO_WAIT)) != NULL) {
```
I did encounter some race condition when combining workqueues with a FIFO.

Excerpt from `#kernel` on Discord ([thread](https://discord.com/channels/720317445772017664/884851987508781066/1344094308755832903)):

> <@andyross> @josuah Unless I'm misunderstanding, that sounds like a huge footgun if your work item and queue entry are 1:1.  What happens if you race and they interleave?  The work queue is already conceptually a FIFO.  A FIFO can be used to implement a main loop for a work-queue-like thread.  Pick one or the other, not both. 🙂

I know this is how the other drivers are implemented, but if interested in both simplifying the BFLB driver and avoid this possible gotcha, one possibility would be to set the `k_work` as a field of `struct udc_bflb_bl61x_ev`, which would be a "piece of work".

This mean you do not need `priv->fifo` anymore.

VynDragon replied:
Well i have *no* clue how this works but i made it work very easily. Can you confirm this is the proper usage?

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+
+struct udc_bflb_bl61x_data {
+	struct k_work work;
+	struct k_fifo fifo;
```
[Ctrl+F #kernel]
As seen below, it might be possible to remove these two fields.

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+
+	if (fifo_idx > 4) {
+		return;
+	}
```
If it is an invalid configuration, as an alternative:
```suggestion
	__ASSERT_NO_MSG(fifo_idx <= 4);
```

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+
+	if (ep_idx > 4 || fifo_idx > 4 || fifo_dir > 2) {
+		return;
+	}
```
In case this was for checking invalid configuration, since only used internally:
```suggestion
	__ASSERT_NO_MSG(ep_idx <= 4);
	__ASSERT_NO_MSG(fifo_idx <= 4);
	__ASSERT_NO_MSG(fifo_dir <= 2);
```

VynDragon replied:
done

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+	sys_write32(tmp, cfg->base + USB_VDMA_F0PS1_OFFSET
+		+ (fifo_idx - 1) * USB_BL61X_FX_X_OFFSET);
+
+	sys_cache_data_flush_and_invd_range(buf, len);
```
It did not seem like causing trouble though!

VynDragon replied:
The expectation here is that the cache lines relevant to the flush will not be read until the vdma is done, which means it doesnt cache anything between that flush and post-vdma read, which means it gives the same result as doing it post-dma. I am not sure it is possible to do that flush after the dma due to how the code is

VynDragon replied:
Well i tried swapping that out for end update and that didnt quit work out, so ill leave it as is.

josuah reviewed (commented):
I wish to do another round of review when I get back home, also run USB3CV compliance tool which will be good at spotting bugs.

Though it sounds like ready for review to me.
Thank you!

josuah commented:
I think you are right, to follow the UDC (USB Device Controller) acronym, `BL61x USB controller device driver` could eventually become `BL61x USB device controller driver`.

VynDragon commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1480 @@
+				   USB_EP_TYPE_CONTROL, 64, 0)) {
+		LOG_ERR("Failed to enable control endpoint");
+		return -EIO;
+	}
```
done

VynDragon commented:
And one other caveat before this is ready, this will need to add a new board, because the current bl61x board ai-m62-12f has NO usb at all going out the module so it cant support USB. Liekly this will be sipeed m0s, but it could also be added to qcc744_evk if it's merged very fast

VynDragon committed:
boards: bflb: Add USB to ai_m62_12f

This is necessary to have a board testing the USB driver,
however the board doesn't have a USB plug for it,
but the pins are available

Signed-off-by: Camille BAUD <mail@massdriver.space>

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1473 @@
+	LOG_DBG("dev %p, ep 0x%02x, event %u",
+		ev->dev, ev->ep_addr, ev->event);
+
+	ep_cfg = udc_get_ep_cfg(ev->dev, ev->ep_addr);
```
If applying the "const everywhere", then this could be refactored like this:

```suggestion
	struct udc_bflb_bl61x_ev *const ev = CONTAINER_OF(item, struct udc_bflb_bl61x_ev, work);
	struct udc_ep_config *const ep_cfg = udc_get_ep_cfg(ev->dev, ev->ep_addr);
	int err = 0;

	LOG_DBG("dev %p, ep 0x%02x, event %u",
		ev->dev, ev->ep_addr, ev->event);
```

VynDragon replied:
i think i got most of it...

josuah replied:
Thanks! If really wanting to convert *all* I've annotated them, but I don't see current drivers being applying this 100% either.

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1473 @@
+
+static enum udc_bus_speed udc_bflb_bl61x_device_speed(const struct device *dev)
+{
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
```
And if so this could also apply to every struct assignment for `data` and `cfg`:
```suggestion
	const struct udc_bflb_bl61x_config *const cfg = dev->config;
```

josuah reviewed (dismissed):
LGTM thank you I did not find anything missing!
I added a few extra notes if wanting to anticipate the review of USB maintainers, not 100% sure this is going to be requested.

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+static void udc_bflb_bl61x_ctrl_dout_start(const struct device *const dev, uint16_t size)
+{
+	struct net_buf *buf;
+	struct udc_ep_config *ep_cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
```
```suggestion
	struct udc_ep_config *const ep_cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+{
+	const struct udc_bflb_bl61x_config *const cfg = dev->config;
+	uint32_t tmp;
+	uint8_t ep_idx = USB_EP_GET_IDX(config->addr);
```
```suggestion
	const uint8_t ep_idx = USB_EP_GET_IDX(config->addr);
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+{
+	const struct udc_bflb_bl61x_config *const cfg = dev->config;
+	uint32_t tmp;
+	uint8_t ep_idx = USB_EP_GET_IDX(config->addr);
```
```suggestion
	const uint8_t ep_idx = USB_EP_GET_IDX(config->addr);
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+	return UDC_BUS_UNKNOWN;
+}
+
+static uint8_t udc_bflb_bl61x_fifo_get_ep(const struct device *const dev, uint8_t fifo)
```
```suggestion
static uint8_t udc_bflb_bl61x_fifo_get_ep(const struct device *const dev, const uint8_t fifo)
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+
+static void udc_bflb_bl61x_ep_set_out_mps(const struct device *const dev,
+					uint8_t ep_idx,
+					uint16_t ep_mps)
```
```suggestion
					const uint8_t ep_idx,
					const uint16_t ep_mps)
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+	sys_write32(tmp, cfg->base + USB_DEV_CXCFE_OFFSET);
+}
+
+static void udc_bflb_bl61x_fifo_reset(const struct device *const dev, uint8_t fifo_idx)
```
```suggestion
static void udc_bflb_bl61x_fifo_reset(const struct device *const dev, const uint8_t fifo_idx)
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+static void udc_bflb_bl61x_ep_setfifo(const struct device *const dev,
+					uint8_t ep_idx,
+					uint8_t fifo_idx,
+					uint8_t ep_dir)
```
```suggestion
					const uint8_t ep_idx,
					const uint8_t fifo_idx,
					const uint8_t ep_dir)
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+static void udc_bflb_bl61x_fifo_setep(const struct device *const dev,
+					uint8_t ep_idx,
+					uint8_t fifo_idx,
+					uint8_t fifo_dir)
```
```suggestion
					const uint8_t ep_idx,
					const uint8_t fifo_idx,
					const uint8_t fifo_dir)
```

josuah commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1475 @@
+}
+
+static void udc_bflb_bl61x_vdma_startwrite_ctrl(const struct device *const dev,
+						uint8_t *data, uint32_t len)
```
```suggestion
						uint8_t *const data, const uint32_t len)
```

josuah reviewed (commented):
I am not sure it's going to be enforced to have all of them marked as `const` (not part of Zephyr coding style, just the style in USB), but in case you wanted a TODO list here we go.

josuah commented:
I think the next step is basically wait that the maintainers' review queue clears out.

If it takes long and "ping" don't help, the workaround is `west patch gh-fetch --pull-request 97750` to maintain this PR as a local patch until it's merged.

VynDragon commented:
> west patch gh-fetch --pull-request 97750

i didnt know about that and manually git fetched and cherry picked PR

jfischer-no commented on drivers/usb/udc/udc_bflb_bl61x.c:
Looking at the next commit I was curious about "It is similar to FOTG210". I started to look at the BL61X manual, and there does not seem to be one, the same appears to be for FOTG210 IP. I see some rumors that the BL61X SoC uses FOTG210 IP. Comparing fotg210-udc.[c,h] from the Linux kernel with your code and usb_v2_reg.h from the bouffalolab HAL, it seems to be FOTG210 USB controller IP. In that case, the right way is to implement platform/vendor independent driver for FOTG210 based controllers and use vendor specific quirks for platform specific parts like clock configuration or pin configuration.

Please use `drivers/usb/common/usb_dwc2_hw.h`, `drivers/usb/udc/udc_dwc2.[c,h]`, and  `drivers/usb/udc/udc_dwc2_vendor_quirks.h` as the reference.
Rename `udc_bflb_bl61x.c` to `udc_fotg210.c` and use udc_fotg210/fotg210 as namespace. Add common header file `drivers/usb/common/usb_fotg210_hw.h`. Use definition from `usb_fotg210_hw.h` in `udc_fotg210.c`. Move BL61X specific platform code to `drivers/usb/udc/udc_fotg210_vendor_quirks.h`.

The problem is where to get register names for `drivers/usb/common/usb_fotg210_hw.h`, ideally they should have similar names as in the official FOTG210 manual/databook. I have no idea where to get it. Faraday Technology seems to be the original author of the Linux driver but these files are GPL-2.0+ only and definitions there can not be copied. So this need to be solved first.

I am also curious where did you get all the info to write this driver?

VynDragon commented:
Converting back to draft because some major bugs have been found and we're struggling to fix them (what with not having docs...).

VynDragon commented:
not stale just still buggered :(

VynDragon commented:
scratch that, it works now, and some people getting full pass:

```❯ sudo testusb -v 512 -D /dev/bus/usb/003/083       
testusb: /dev/bus/usb/003/083 may see only control tests
high speed      /dev/bus/usb/003/083    0
/dev/bus/usb/003/083 test 0,    0.000017 secs
/dev/bus/usb/003/083 test 1,    0.386155 secs
/dev/bus/usb/003/083 test 2,    0.416616 secs
/dev/bus/usb/003/083 test 3,    0.231272 secs
/dev/bus/usb/003/083 test 4,    0.208546 secs
/dev/bus/usb/003/083 test 5,    9.963698 secs
/dev/bus/usb/003/083 test 6,   13.453021 secs
/dev/bus/usb/003/083 test 7,    5.152373 secs
/dev/bus/usb/003/083 test 8,    7.008731 secs
/dev/bus/usb/003/083 test 9,    1.155443 secs
/dev/bus/usb/003/083 test 10,    5.255097 secs
/dev/bus/usb/003/083 test 11,    3.122939 secs
/dev/bus/usb/003/083 test 12,    2.982593 secs
/dev/bus/usb/003/083 test 13,    3.860037 secs
/dev/bus/usb/003/083 test 14,    0.745634 secs
/dev/bus/usb/003/083 test 17,    0.384816 secs
/dev/bus/usb/003/083 test 18,    0.416489 secs
/dev/bus/usb/003/083 test 19,    0.378912 secs
/dev/bus/usb/003/083 test 20,    0.416622 secs
/dev/bus/usb/003/083 test 21,    0.740386 secs
/dev/bus/usb/003/083 test 24,   11.761522 secs
/dev/bus/usb/003/083 test 27,    9.819874 secs
/dev/bus/usb/003/083 test 28,   13.333067 secs
/dev/bus/usb/003/083 test 29,    1.055241 secs
```
updating PR soon...

Hiihtaja77 commented on drivers/usb/udc/udc_bflb_bl61x.c:
```diff
@@ -0,0 +1,1556 @@
+	static const struct udc_bflb_bl61x_config udc_bflb_bl61x_config_##n = {	\
+		.base = DT_INST_REG_ADDR(n),					\
+		.num_of_eps = DT_INST_PROP(n, num_bidir_endpoints),		\
+		.ep_cfg_in = ep_cfg_out,					\
```
does .ep_cfg_in = ep_cfg_out,  in to out mapping is OK ?
Or it is dosn't matter ?
